plugin.program.AML/resources/main.py

```diff
@@ -0,0 +1,6751 @@
+# -*- coding: utf-8 -*-
+#
+# Advanced MAME Launcher main script file
+#
+
+# Copyright (c) 2016-2018 Wintermute0110 <wintermute0110@gmail.com>
+#
+# This program is free software; you can redistribute it and/or modify
+# it under the terms of the GNU General Public License as published by
+# the Free Software Foundation; version 2 of the License.
+#
+# This program is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+# GNU General Public License for more details.
+
+# --- Python standard library ---
+# Division operator: https://www.python.org/dev/peps/pep-0238/
+from __future__ import unicode_literals
+from __future__ import division
+import os
+import urlparse
+import subprocess
+import copy
+from datetime import datetime
+
+# --- Kodi stuff ---
+import xbmc, xbmcgui, xbmcplugin, xbmcaddon
+
+# --- Modules/packages in this plugin ---
+# Addon module dependencies:
+#   main <-- mame <-- disk_IO <-- assets, utils, utils_kodi, constants
+#   ReaderPDF <-- utils, utils_kodi
+#   filters <- mame <-- utils, utils_kodi
+#   manuals <- utils, utils_kodi
+from constants import *
+from assets import *
+from utils import *
+from utils_kodi import *
+from disk_IO import *
+from mame import *
+from filters import *
+from manuals import *
+
+# --- Addon object (used to access settings) ---
+__addon__         = xbmcaddon.Addon()
+__addon_id__      = __addon__.getAddonInfo('id').decode('utf-8')
+__addon_name__    = __addon__.getAddonInfo('name').decode('utf-8')
+__addon_version__ = __addon__.getAddonInfo('version').decode('utf-8')
+__addon_author__  = __addon__.getAddonInfo('author').decode('utf-8')
+__addon_profile__ = __addon__.getAddonInfo('profile').decode('utf-8')
+__addon_type__    = __addon__.getAddonInfo('type').decode('utf-8')
+
+# --- Addon paths and constant definition ---
+# _PATH is a filename | _DIR is a directory
+ADDONS_DATA_DIR      = FileName('special://profile/addon_data')
+PLUGIN_DATA_DIR      = ADDONS_DATA_DIR.pjoin(__addon_id__)
```

Comment:
The FileName() class is designed to deal with all file operations and abstracts URLs, normal files, files in network paths, etc. Currently the FileName() class uses the Python API but in Leia/Retroplayer it will use Kodi VFS so users can play ROMs in networks shares. I think is would be a bad idea to tamper with that piece of code because it will affect all the database files.